pkg/gadgets/run/tracer/run.go

```diff
@@ -754,3 +680,185 @@ func (g *GadgetDesc) EventPrototype() any {
 	gadgetregistry.Register(&GadgetDesc{})
 }
+
+// functions to send columns from server to client
+func typeFromBTF(typ btf.Type) *types.Type {
+	switch typedMember := typ.(type) {
+	case *btf.Array:
+		arrType := simpleTypeFromBTF(typedMember.Type)
+		if arrType == nil {
+			return nil
+		}
+		return &types.Type{
+			Kind:           types.KindArray,
+			ArrayNElements: int(typedMember.Nelems),
+			ArrayKind:      arrType.Kind,
+		}
+	default:
+		return simpleTypeFromBTF(typ)
+	}
+}
+
+func simpleTypeFromBTF(typ btf.Type) *types.Type {
+	switch typedMember := typ.(type) {
+	case *btf.Int:
+		switch typedMember.Encoding {
+		case btf.Signed:
+			switch typedMember.Size {
+			case 1:
+				return &types.Type{Kind: types.KindInt8}
+			case 2:
+				return &types.Type{Kind: types.KindInt16}
+			case 4:
+				return &types.Type{Kind: types.KindInt32}
+			case 8:
+				return &types.Type{Kind: types.KindInt64}
+			}
+		case btf.Unsigned:
+			switch typedMember.Size {
+			case 1:
+				return &types.Type{Kind: types.KindUint8}
+			case 2:
+				return &types.Type{Kind: types.KindUint16}
+			case 4:
+				return &types.Type{Kind: types.KindUint32}
+			case 8:
+				return &types.Type{Kind: types.KindUint64}
+			}
+		case btf.Bool:
+			return &types.Type{Kind: types.KindBool}
+		case btf.Char:
+			return &types.Type{Kind: types.KindUint8}
+		}
+	case *btf.Float:
+		switch typedMember.Size {
+		case 4:
+			return &types.Type{Kind: types.KindFloat32}
+		case 8:
+			return &types.Type{Kind: types.KindFloat64}
+		}
+	case *btf.Typedef:
+		typ, _ := getUnderlyingType(typedMember)
+		return simpleTypeFromBTF(typ)
+	case *btf.Enum:
+		if typedMember.Signed {
+			switch typedMember.Size {
+			case 1:
+				return &types.Type{Kind: types.KindInt8}
+			case 2:
+				return &types.Type{Kind: types.KindInt16}
+			case 4:
+				return &types.Type{Kind: types.KindInt32}
+			case 8:
+				return &types.Type{Kind: types.KindInt64}
+			}
+		}
+
+		switch typedMember.Size {
+		case 1:
+			return &types.Type{Kind: types.KindUint8}
+		case 2:
+			return &types.Type{Kind: types.KindUint16}
+		case 4:
+			return &types.Type{Kind: types.KindUint32}
+		case 8:
+			return &types.Type{Kind: types.KindUint64}
+		}
+	}
+
+	return nil
+}
+
+func calculateColumnsForClient(
+	eventFactory *types.EventFactory,
+	gadgetMetadata *types.GadgetMetadata,
+	progContent []byte,
+) ([]types.ColumnDesc, error) {
+	eventType, err := getEventTypeBTF(progContent, gadgetMetadata)
+	if err != nil {
+		return nil, fmt.Errorf("getting value struct: %w", err)
+	}
+
+	colNames := map[string]struct{}{}
+
+	eventStruct, ok := gadgetMetadata.Structs[eventType.Name]
+	if !ok {
+		return nil, fmt.Errorf("struct %s not found in gadget metadata", eventType.Name)
+	}
+
+	for _, field := range eventStruct.Fields {
+		colNames[field.Name] = struct{}{}
+	}
+
+	columns := []types.ColumnDesc{}
+
+	for _, member := range eventType.Members {
+		member := member
+
+		_, ok := colNames[member.Name]
+		if !ok {
+			continue
```

Comment:
Maybe add debug message that that member was skipped because it is not in the gadget metadata